File: 2016/09.cs
namespace _2016;

public abstract class _09
{
    private static readonly string Data;

    static _09() => Data = string.Concat(Task.Run(() => Utils.GetInputData(9)).Result
        .Split('\n', StringSplitOptions.RemoveEmptyEntries)).Replace(" ", "");

    private static long CalculateDecompressedLength(string input, int start, int end, bool recursive)
    {
        long decomLen = 0;

        for (var i = start; i < end;)
        {
            if (input[i] == '(')
            {
                var markerEnd = input.IndexOf(')', i);
                if (markerEnd == -1) break;

                var marker = input.Substring(i + 1, markerEnd - i - 1);
                var parts = marker.Split('x');
                if (parts.Length != 2 || !int.TryParse(parts[0], out var seqLen) ||
                    !int.TryParse(parts[1], out var repCount)) break;

                var subStart = markerEnd + 1;
                var subEnd = subStart + seqLen;

                if (recursive) decomLen += repCount * CalculateDecompressedLength(input, subStart, subEnd, true);
                else decomLen += seqLen * repCount;

                i = subEnd;
            }
            else
            {
                decomLen++;
                i++;
            }
        }

        return decomLen;
    }

    private static long SolvePartOne() => CalculateDecompressedLength(Data, 0, Data.Length, false);

    private static long SolvePartTwo() => CalculateDecompressedLength(Data, 0, Data.Length, true);

    public static void Run()
    {
        Console.WriteLine($"Part One: {SolvePartOne()}");
        Console.WriteLine($"Part Two: {SolvePartTwo()}");
    }
}

File: 2017/09.fs
namespace _2017

open System

module _09 =
    let Data = (Utils.GetInputData 9).Split('\n', StringSplitOptions.RemoveEmptyEntries)

    let processStream processChar =
        let stream = Data[0]
        let mutable inGarbage = false
        let mutable skipNext = false
        for c in stream do
            match (skipNext, inGarbage, c) with
            | true, _, _ -> skipNext <- false
            | _, _, '!' -> skipNext <- true
            | _, true, '>' -> inGarbage <- false
            | _, true, _ -> processChar (c, inGarbage)
            | _, _, '<' -> inGarbage <- true
            | _, _, _ -> processChar (c, inGarbage)

    let solvePartOne() =
        let mutable score = 0
        let mutable depth = 0
        processStream (fun (c, inGarbage) ->
            if not inGarbage then
                match c with
                | '{' ->
                    depth <- depth + 1
                    score <- score + depth
                | '}' -> depth <- depth - 1
                | _ -> ())
        score

    let solvePartTwo() =
        let mutable garbageCount = 0
        processStream (fun (_, inGarbage) -> if inGarbage then garbageCount <- garbageCount + 1)
        garbageCount

    let Run() =
        printfn $"Part One: {solvePartOne()}"
        printfn $"Part Two: {solvePartTwo()}"